*Последние изменения внесены: `21.03.2023`*

# Requests в Python – Примеры выполнения HTTP запросов

* `Библиотека requests` является стандартным инструментом для составления **HTTP-запросов** в **Python**. 
* Простой и аккуратный **API** значительно облегчает трудоемкий процесс создания запросов.

Стоит взять на заметку сайт **httpbin.org**. 
   * Это чрезвычайно полезный ресурс, созданный человеком, который внедрил использование `requests` – *Кеннетом Рейтцом*.
   * Данный сервис предназначен для тестовых запросов. 
   * Здесь можно составить пробный запрос и получить ответ с требуемой информацией. 

### HTTP Request (запрос) и структура URL

Типовая структура URL:

```http://site.com/path/to/resource?param1=value1*&*param2=value2```


HTTP запрос состоит из:
* **query string** - _строка запроса_, содержащая:
    * **method**  - _метод запроса_ 
    * **URL** или **URI** - _адресная строка запроса_, 
    * **params** - _параметры запроса_, используются для фильтрации информации, которую мы хотим получить.
        * _Знак `?` отделяет строку URL от параметров._
        * _Парметры идут в паре `ключ=значение`_
        * _Знак `&` - амперсанд отделяет одни параметры от других._
* **headers** - _заголовок_ - это служебная часть содержащая информацию о запросе, например:
    * авторизационные данные пользователя
    * версию браузера
    * выставленные cookie
    * поддерживаемые форматы сжатия данных
    * и т.д.
* **body** - _тело запроса_

Простейший запрос может не содержать ни заголовков, ни метода, ни параметров. Необходим только метод и адресная строка запроса.

## 1. Python установка `библиотеки requests`

Для начала работы потребуется установить библиотеку requests. Для этого используется следующая команда:

`pip install requests`


Сразу после установки requests можно полноценно использовать в приложении. Импорт requests производится следующим образом:
```python
import requests
```

## 2. Python библиотека `Requests` метод `GET`

* **GET** является одним из самых популярных **HTTP методов**. 
* Метод **GET** указывает на то, что происходит попытка извлечь данные из определенного ресурса. 
* Для того, чтобы выполнить запрос **GET**, используется `requests.get()`.
* Метод **GET** - относится к безопасным методам, так как не выполняет никаких дейсвтий на сервере.

> Для проверки работы команд будем выполнять запрос в отношении **Root REST API** на **GitHub**.`

Выполним **запрос GET** в отношении указанного в скобках URL (вызовем `метод get()`):
```python
request.get('https://api.github.com')
```

* Если никакие python ошибки не возникло – первый запрос успешно выполнен. Далее будет рассмотрен ответ на данный запрос, который можно получить при помощи объекта `Response`.

## 3. Объект `Response` получение ответа на запрос в Python

* `Response` представляет собой довольно мощный объект для анализа результатов запроса. 
* В качестве примера будет использован предыдущий запрос, только на этот раз результат будет представлен в виде переменной.
* Таким образом, получится лучше изучить его атрибуты и особенности использования.

```python
import requests

response = requests.get('https://api.github.com')
```

* В данном примере при помощи `get(`) захватывается определенное значение, что является частью объекта `Response`, и помещается в переменную под названием **response**. 
* Теперь можно использовать переменную **response** для того, чтобы изучить данные, которые были получены в результате запроса **GET**.

## 4. HTTP коды состояний

* Самыми первыми данными, которые будут получены через `Response`, будут **коды состояния**. 
* **Коды состояния** сообщают о статусе запроса.
* Например, статус **200 OK** значит, что запрос успешно выполнен. А вот статус **404 NOT FOUND** говорит о том, что запрашиваемый ресурс не был найден. 
* Существует множество других статусных кодов, которые могут сообщить важную информацию, связанную с запросом.
* Интересный сайт с расшифровкой кодов: https://http.cat/

**HTTP коды**

Первая цифра обозначает семантику кода.
* 1XX — Информационные
* 2XX — Успешный вызов (пример: 200 – ок, 201 – создано)
* 3XX — Перенаправление
* 4XX — Ошибка на стороне клиента (пример: 404 – не найдено, 403 – недостаточно прав) 
* 5XX — Ошибка на стороне сервера

### `.status_code`

* Используя `.status_code`, можно увидеть код состояния, который возвращается с сервера:

In [13]:
import requests

response = requests.get('https://api.github.com')
print(response.status_code)

200


* `.status_code` вернул значение **200**. Это значит, что запрос был выполнен успешно, а сервер ответил, отобразив запрашиваемую информацию.

В некоторых случаях необходимо использовать полученную информацию для написания программного кода.

Например:

In [12]:
import requests

response = requests.get('https://api.github.com')

if response.status_code == 200:
    print('Success!')
elif response.status_code == 404:
    print('Not Found.')

Success!


* В таком случае, если с сервера будет получен код состояния **200**, тогда программа выведет значение **Success!**. Однако, если от сервера поступит код **404**, тогда программа выведет значение **Not Found**.

 Если использовать `Response` в условных конструкциях, то при получении кода состояния в промежутке от **200** до **400**, будет выведено значение **True**. В противном случае отобразится значение **False**:

In [11]:
import requests

response = requests.get('https://api.github.com')

if response:
    print('Success!')
else:
    print('An error has occurred.')

Success!


* Стоит иметь в виду, что данный способ не проверяет, имеет ли статусный код точное значение **200**. 
* Причина заключается в том, что другие коды в промежутке от **200** до **400**, например, **204 NO CONTENT** и **304 NOT MODIFIED**, также считаются успешными в случае, если они могут предоставить действительный ответ.
* К примеру, код состояния **204** говорит о том, что ответ успешно получен, однако в полученном объекте нет содержимого.
* Можно сказать, что для оптимально эффективного использования способа необходимо убедиться, что начальный запрос был успешно выполнен. 
* Требуется изучить код состояния и в случае необходимости произвести необходимые поправки, которые будут зависеть от значения полученного кода.

### `.raise_for_status()` и `HTTPError`

* Если при использовании оператора **if** вы не хотите проверять код состояния, можно расширить диапазон исключений для неудачных результатов запроса. 
* Это можно сделать при помощи использования `.raise_for_status()`

In [10]:
import requests
from requests.exceptions import HTTPError
 
for url in ['https://api.github.com']:
    try:
        response = requests.get(url)
 
        # если ответ успешен, исключения задействованы не будут
        response.raise_for_status()
    except HTTPError as http_err:
        print(f'HTTP error occurred: {http_err}')  # Python 3.6
    except Exception as err:
        print(f'Other error occurred: {err}')  # Python 3.6
    else:
        print('Success!')

Success!


* В случае вызова исключений через `.raise_for_status()` к некоторым кодам состояния применяется `HTTPError`. 
* Когда код состояния показывает, что запрос успешно выполнен, программа продолжает работу без применения политики исключений.

**Подыто́жим:**
* Анализ способов использования кодов состояния, полученных с сервера, является неплохим стартом для изучения `requests`.
* Тем не менее, при создании запроса **GET**, значение кода состояния является не самой важной информацией, которую хочет получить программист. 
* Обычно запрос производится для извлечения более содержательной информации. 
* В дальнейшем будет показано, как добраться до актуальных данных, которые сервер высылает отправителю в ответ на запрос.

## 5. Получить содержимое страницы в `Requests`

### `payload`

* Ответ на запрос **GET** содержит информацию. 
* Она находится в теле сообщения и называется **пейлоад** (`payload`). 
* Используя атрибуты и методы библиотеки `Response`, можно получить **пейлоад** в различных форматах.



### `.content` ,  `.text`  ,  `.encoding=" "`  ,  `json()`

* Для того, чтобы получить содержимое запроса в байтах, необходимо использовать `.content`.
* Использование `.content` обеспечивает доступ к чистым байтам ответного пейлоада, то есть к любым данным в теле запроса.

In [18]:
import requests

response = requests.get('https://api.github.com')
response.content

b'{"current_user_url":"https://api.github.com/user","current_user_authorizations_html_url":"https://github.com/settings/connections/applications{/client_id}","authorizations_url":"https://api.github.com/authorizations","code_search_url":"https://api.github.com/search/code?q={query}{&page,per_page,sort,order}","commit_search_url":"https://api.github.com/search/commits?q={query}{&page,per_page,sort,order}","emails_url":"https://api.github.com/user/emails","emojis_url":"https://api.github.com/emojis","events_url":"https://api.github.com/events","feeds_url":"https://api.github.com/feeds","followers_url":"https://api.github.com/user/followers","following_url":"https://api.github.com/user/following{/target}","gists_url":"https://api.github.com/gists{/gist_id}","hub_url":"https://api.github.com/hub","issue_search_url":"https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}","issues_url":"https://api.github.com/issues","keys_url":"https://api.github.com/user/keys","label_sea

* Однако, зачастую требуется конвертировать полученную информацию в строку в кодировке UTF-8. response делает это при помощи `.text`.

In [19]:
import requests

response = requests.get('https://api.github.com')
response.text

'{"current_user_url":"https://api.github.com/user","current_user_authorizations_html_url":"https://github.com/settings/connections/applications{/client_id}","authorizations_url":"https://api.github.com/authorizations","code_search_url":"https://api.github.com/search/code?q={query}{&page,per_page,sort,order}","commit_search_url":"https://api.github.com/search/commits?q={query}{&page,per_page,sort,order}","emails_url":"https://api.github.com/user/emails","emojis_url":"https://api.github.com/emojis","events_url":"https://api.github.com/events","feeds_url":"https://api.github.com/feeds","followers_url":"https://api.github.com/user/followers","following_url":"https://api.github.com/user/following{/target}","gists_url":"https://api.github.com/gists{/gist_id}","hub_url":"https://api.github.com/hub","issue_search_url":"https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}","issues_url":"https://api.github.com/issues","keys_url":"https://api.github.com/user/keys","label_sear

* Декодирование байтов в строку требует наличия определенной модели кодировки. 
* По умолчанию `requests` попытается узнать текущую кодировку, ориентируясь по заголовкам **HTTP**. 
* Указать необходимую кодировку можно при помощи добавления `.encoding` перед `.text`.

In [40]:
import requests

response = requests.get('https://api.github.com')
response.encoding = 'utf-8' 
response.text

'{\n  "current_user_url": "https://api.github.com/user",\n  "current_user_authorizations_html_url": "https://github.com/settings/connections/applications{/client_id}",\n  "authorizations_url": "https://api.github.com/authorizations",\n  "code_search_url": "https://api.github.com/search/code?q={query}{&page,per_page,sort,order}",\n  "commit_search_url": "https://api.github.com/search/commits?q={query}{&page,per_page,sort,order}",\n  "emails_url": "https://api.github.com/user/emails",\n  "emojis_url": "https://api.github.com/emojis",\n  "events_url": "https://api.github.com/events",\n  "feeds_url": "https://api.github.com/feeds",\n  "followers_url": "https://api.github.com/user/followers",\n  "following_url": "https://api.github.com/user/following{/target}",\n  "gists_url": "https://api.github.com/gists{/gist_id}",\n  "hub_url": "https://api.github.com/hub",\n  "issue_search_url": "https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}",\n  "issues_url": "https://api.g

* Если присмотреться к ответу, можно заметить, что его содержимое является сериализированным **JSON** контентом.
* Воспользовавшись словарем, можно взять полученные из `.text` строки **str** и провести с ними обратную сериализацию при помощи использования `json.loads()`. 
* Есть и более простой способ, который требует применения `.json()`.

In [41]:
import requests

response = requests.get('https://api.github.com')
response.json()

{'current_user_url': 'https://api.github.com/user',
 'current_user_authorizations_html_url': 'https://github.com/settings/connections/applications{/client_id}',
 'authorizations_url': 'https://api.github.com/authorizations',
 'code_search_url': 'https://api.github.com/search/code?q={query}{&page,per_page,sort,order}',
 'commit_search_url': 'https://api.github.com/search/commits?q={query}{&page,per_page,sort,order}',
 'emails_url': 'https://api.github.com/user/emails',
 'emojis_url': 'https://api.github.com/emojis',
 'events_url': 'https://api.github.com/events',
 'feeds_url': 'https://api.github.com/feeds',
 'followers_url': 'https://api.github.com/user/followers',
 'following_url': 'https://api.github.com/user/following{/target}',
 'gists_url': 'https://api.github.com/gists{/gist_id}',
 'hub_url': 'https://api.github.com/hub',
 'issue_search_url': 'https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}',
 'issues_url': 'https://api.github.com/issues',
 'keys_url': '

* Тип полученного значения из `.json()`, является словарем. 
* Это значит, что доступ к его содержимому можно получить по ключу:

In [27]:
response.json()['current_user_url']

'https://api.github.com/user'

> **Коды состояния** и **тело сообщения** предоставляют огромный диапазон возможностей. Однако, для их оптимального использования требуется изучить метаданные и заголовки **HTTP**.

* Далее мы рассмотрим **параметры запроса** и **заголовки**.
* Чтобы передать параметры и заголовки в запросе, лучше использовать **именованные аргументы**:

```python
import requests
ur = "https://httpbin.org/get
params = {"foo": "bar", "message": "hello"}
headers = {"Authorization": "secret-token-123"}
resp = requests.get(url, params=params, headers=headers)
```
**Важно:**
* Конкретные значения параметров и заголовков зависят от сервера, к которому происходит обращение. 
* Для того чтобы узнать точные значения параметров, нужно читать документацию или проверять опытным путем.


## 6. HTTP заголовки в `Requests`

* **HTTP** заголовки ответов на запрос могут предоставить определенную полезную информацию. 
* Это может быть тип содержимого ответного **пейлоада**, а также ограничение по времени для **кеширования ответа**. 
* Для просмотра **HTTP** заголовков загляните в атрибут `.headers.`



### `.headers`

In [32]:
import requests

response = requests.get('https://api.github.com')
response.headers

{'Server': 'GitHub.com', 'Date': 'Sat, 11 Mar 2023 12:27:52 GMT', 'Cache-Control': 'public, max-age=60, s-maxage=60', 'Vary': 'Accept, Accept-Encoding, Accept, X-Requested-With', 'ETag': '"4f825cc84e1c733059d46e76e6df9db557ae5254f9625dfe8e1b09499c449438"', 'x-github-api-version-selected': '2022-11-28', 'Access-Control-Expose-Headers': 'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=31536000; includeSubdomains; preload', 'X-Frame-Options': 'deny', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '0', 'Referrer-Policy': 'origin-when-cross-origin, strict-origin-when-cross-origin', 'Content-Security-Policy': "default-src 'none'", 'Content-Type': 'application/json; charset=ut

* `.headers` возвращает словарь, что позволяет получить доступ к значению заголовка **HTTP** по ключу. 
* Например, для просмотра типа содержимого ответного **пейлоада**, требуется использовать ключ словаря **'Content-Type'**:

In [31]:
import requests

response = requests.get('https://api.github.com')
response.headers['Content-Type']

'application/json; charset=utf-8'

* У объектов словарей в качестве заголовков есть своим особенности. 
* Специфика **HTTP** предполагает, что заголовки не чувствительны к регистру.
* Это значит, что при получении доступа к заголовкам можно не беспокоится о том, использованы строчным или прописные буквы.
* При использовании ключей **'content-type'** и **'Content-Type'** результат будет получен один и тот же.



**Подыто́жим:**
* Это была основная информация, требуемая для работы с `Response`. 
* Были задействованы главные атрибуты и методы, а также представлены примеры их использования. 
* В дальнейшем будет показано, как изменится ответ после настройки **запроса GET**.

## 7. Python `Requests` параметры запроса (Нихрена не понятно, но очень интересно!)

* Наиболее простым способом настроить **запрос GET** является передача значений через параметры строки запроса в URL.
* При использовании метода `get()`, данные передаются в `params`. 
* Например, для того, чтобы посмотреть на библиотеку `requests` можно использовать **Search API** на **GitHub** :

In [44]:
import requests
 
# Поиск местонахождения для запросов на GitHub
response = requests.get(
    'https://api.github.com/search/repositories',
    params={'q': 'requests+language:python'},
)
 
# Анализ некоторых атрибутов местонахождения запросов
json_response = response.json()
repository = json_response['items'][0]
pprint(repository)
print('=========================================')
print(f'Repository name: {repository["name"]}')  # Python 3.6+
print(f'Repository description: {repository["description"]}')  # Python 3.6+

{'allow_forking': True,
 'archive_url': 'https://api.github.com/repos/spyoungtech/grequests/{archive_format}{/ref}',
 'archived': False,
 'assignees_url': 'https://api.github.com/repos/spyoungtech/grequests/assignees{/user}',
 'blobs_url': 'https://api.github.com/repos/spyoungtech/grequests/git/blobs{/sha}',
 'branches_url': 'https://api.github.com/repos/spyoungtech/grequests/branches{/branch}',
 'clone_url': 'https://github.com/spyoungtech/grequests.git',
 'collaborators_url': 'https://api.github.com/repos/spyoungtech/grequests/collaborators{/collaborator}',
 'comments_url': 'https://api.github.com/repos/spyoungtech/grequests/comments{/number}',
 'commits_url': 'https://api.github.com/repos/spyoungtech/grequests/commits{/sha}',
 'compare_url': 'https://api.github.com/repos/spyoungtech/grequests/compare/{base}...{head}',
 'contents_url': 'https://api.github.com/repos/spyoungtech/grequests/contents/{+path}',
 'contributors_url': 'https://api.github.com/repos/spyoungtech/grequests/contri

* Передавая словарь **{'q': 'requests+language:python'}** в параметр `params`, который является частью `.get()`, можно изменить ответ, что был получен при использовании **Search API**.

* Можно передать параметры в `get()` в форме **словаря**, как было показано выше. Также можно использовать **список кортежей** :

In [38]:
requests.get(
            'https://api.github.com/search/repositories',
            params=[('q', 'requests+language:python')],
            )

<Response [200]>

* Также можно передать значение в байтах :

In [39]:
requests.get(
            'https://api.github.com/search/repositories',
            params=b'q=requests+language:python',
            )

<Response [200]>

* Строки запроса полезны для уточнения параметров в **запросах GET**. 
* Также можно настроить запросы при помощи добавления или изменения заголовков отправленных сообщений.

## 8. Настройка HTTP заголовка запроса (`headers`) (Нихрена не понятно, но очень интересно!)

* Для изменения **HTTP** заголовка требуется передать словарь данного **HTTP** заголовка в `get()` при помощи использования параметра `headers`. 
* Например, можно изменить предыдущий поисковой запрос, подсветив совпадения в результате. 
* Для этого в заголовке `Accept` медиа тип уточняется при помощи `text-match`.

In [45]:
import requests
 
response = requests.get(
    'https://api.github.com/search/repositories',
    params={'q': 'requests+language:python'},
    headers={'Accept': 'application/vnd.github.v3.text-match+json'},
)
 
# просмотр нового массива `text-matches` с предоставленными данными
# о поиске в пределах результатов
json_response = response.json()
repository = json_response['items'][0]
pprint(repository)
print('=========================================')
print(f'Text matches: {repository["text_matches"]}')

{'allow_forking': True,
 'archive_url': 'https://api.github.com/repos/spyoungtech/grequests/{archive_format}{/ref}',
 'archived': False,
 'assignees_url': 'https://api.github.com/repos/spyoungtech/grequests/assignees{/user}',
 'blobs_url': 'https://api.github.com/repos/spyoungtech/grequests/git/blobs{/sha}',
 'branches_url': 'https://api.github.com/repos/spyoungtech/grequests/branches{/branch}',
 'clone_url': 'https://github.com/spyoungtech/grequests.git',
 'collaborators_url': 'https://api.github.com/repos/spyoungtech/grequests/collaborators{/collaborator}',
 'comments_url': 'https://api.github.com/repos/spyoungtech/grequests/comments{/number}',
 'commits_url': 'https://api.github.com/repos/spyoungtech/grequests/commits{/sha}',
 'compare_url': 'https://api.github.com/repos/spyoungtech/grequests/compare/{base}...{head}',
 'contents_url': 'https://api.github.com/repos/spyoungtech/grequests/contents/{+path}',
 'contributors_url': 'https://api.github.com/repos/spyoungtech/grequests/contri

* Заголовок `Accept` сообщает серверу о типах контента, который можно использовать в рассматриваемом приложении. 
* Здесь подразумевается, что все совпадения будут подсвечены, для чего в заголовке используется значение **application/vnd.github.v3.text-match+json**. 
* Это уникальный заголовок `Accept` для **GitHub**. 
* В данном случае содержимое представлено в специальном **JSON** формате.

Перед более глубоким изучением способов редактирования запросов, будет не лишним остановиться на некоторых других методах **HTTP**.

## 9. Примеры HTTP методов в `Requests` ( !!!Надо дорабатывать!!! )

* Помимо **GET**, большой популярностью пользуются такие методы, как:
    * **Безопасные методы для получения данных с сервера:**
        * `GET` - рассмотрен выше. Один из самыъ частостречаемых методов.
        * `HEAD` - в отличии от метода `get` выдает не всю информацию, а только заголовки ресурсов.
        
    * **Методы с "телом" для отправки данных на сервер:**
        * `POST` - _Один из самых частовстречаемых методов. Создает на сервере новую запись или ресурс с помощью данных, которые у нас лежат в теле запроса._ (_**примеры запроса:** регистрация нового пользователя на сайте или создание заказа товаров из корзины._)
        * `PATCH` - Позволяет частично редактировать данные отправленные методом `POST` (_**примеры запроса:** изменение кол-ва товара в корзине или частичная замена_)
        * `PUT` - Полностью изменяет или заменяет данные отправленные методом `POST` (_**примеры запроса:** полное удаление содержимого нашего заказа, на сайте остается только номер и дата заказа для статистики._)
    * **Методы отмены:**
        * `DELETE` - полностью удаляет наш ресурс или данные с сервера. Результат срабатывания не гарантирован, зависит от настроек доступа к серверу.
        * `TRACE` - возвращает обратно наш запрос к серверу, что-то вроде пинга.
        * `OPTIONS` - возвращает информацию о том, что мы можем делать на этом сервисе или по этому адресу запроса. Доступные нам действия на сервере. Своеобразная функция **help**
* Для каждого из этих методов существует своя сигнатура, которая очень похожа на метод `get()`.

В примерах ниже, каждая функция создает запрос к **httpbin сервису**, используя при этом ответный **HTTP метод**. 

In [51]:
requests.post('https://httpbin.org/post', data={'key':'value'})

<Response [200]>

In [52]:
requests.put('https://httpbin.org/put', data={'key':'value'})

<Response [200]>

In [53]:

requests.delete('https://httpbin.org/delete')

<Response [200]>

In [54]:
requests.head('https://httpbin.org/get')

<Response [200]>

In [55]:
requests.patch('https://httpbin.org/patch', data={'key':'value'})

<Response [200]>

In [56]:
requests.options('https://httpbin.org/get')

<Response [200]>

* При использовании каждого из данных методов в `Response` могут быть возвращены **заголовки**, **тело запроса**, **коды состояния** и многие другие аспекты.

## 10. Python `Requests` тело сообщения

* В соответствии со спецификацией **HTTP** запросы `POST`, `PUT` и `PATCH` передают информацию через тело сообщения, а не через параметры строки запроса. 
* Используя `requests`, можно передать данные в параметр `data`.
* В свою очередь `data` использует **словарь**, **список кортежей**, **байтов** или **объект файла**.
* Это особенно важно, так как может возникнуть необходимость адаптации отправляемых с запросом данных в соответствии с определенными параметрами сервера.
* В том случае, если требуется отравить данные **JSON**, можно использовать параметр `json`. 
* При передачи данных **JSON** через `json()`, `requests` произведет **сериализацию** данных и добавит правильный **Content-Type** заголовок.

* Стоит взять на заметку сайт **httpbin.org**. 
* Это чрезвычайно полезный ресурс, созданный человеком, который внедрил использование `requests` – Кеннетом Рейтцом. 
* Данный сервис предназначен для тестовых запросов. 
* Здесь можно составить пробный запрос и получить ответ с требуемой информацией. 

В качестве примера рассмотрим базовый запрос с использованием `POST`:

In [64]:
response = requests.post('https://httpbin.org/post', json={'key':'value'})
json_response = response.json()
pprint(json_response)
json_response['data']

{'args': {},
 'data': '{"key": "value"}',
 'files': {},
 'form': {},
 'headers': {'Accept': '*/*',
             'Accept-Encoding': 'gzip, deflate, br',
             'Content-Length': '16',
             'Content-Type': 'application/json',
             'Host': 'httpbin.org',
             'User-Agent': 'python-requests/2.28.1',
             'X-Amzn-Trace-Id': 'Root=1-640c9f37-64f8d2ea63e8d2962d0d3189'},
 'json': {'key': 'value'},
 'origin': '5.34.70.54',
 'url': 'https://httpbin.org/post'}


'{"key": "value"}'

In [63]:
json_response['headers']['Content-Type']

'application/json'

* Здесь видно, что сервер получил данные и **HTTP** заголовки, отправленные вместе с запросом. 
* `requests` также предоставляет информацию в форме `PreparedRequest`.

### `Загрузка файлов`

Чтобы передавать файл, его нужно открыть в байтовом режиме и передать объект файла в параметр `files`:
```python
import requests
with open('gifs/your_file.mp4', 'rb') as f:
    resp = requests.post('http://httpbin.org/post', files={"file": f})
```

## 11. Python `Requests` анализ запроса (Надо углубляться!!!)

* При составлении запроса стоит иметь в виду, что перед его фактической отправкой на целевой сервер библиотека `requests` выполняет определенную подготовку. 
* Подготовка запроса включает в себя такие вещи, как проверка заголовков и сериализация содержимого **JSON**.

### `.request`

* Если открыть `.request`, можно просмотреть **PreparedRequest** (*пер. "подготовленный запрос"*).

In [66]:
response = requests.post('https://httpbin.org/post', json={'key':'value'})
pprint(response)
response.request.headers['Content-Type']

<Response [200]>


'application/json'

In [67]:
response.request.url

'https://httpbin.org/post'

In [68]:
response.request.body

b'{"key": "value"}'

* Проверка **PreparedRequest** открывает доступ ко всей информации о выполняемом запросе. 
* Это может быть пейлоад, URL, заголовки, аутентификация и многое другое.

* У всех описанных ранее типов запросов была одна общая черта – они представляли собой **неаутентифицированные запросы к публичным API**. 
* Однако, подобающее большинство служб, с которыми может столкнуться пользователь, **запрашивают аутентификацию**.

## 12. Python Requests аутентификация HTTP AUTH

* Аутентификация помогает сервису понять, кто вы. 
* Как правило, вы предоставляете свои учетные данные на сервер, передавая данные через заголовок Authorization или пользовательский заголовок, определенной службы. 
* Все функции запроса, которые вы видели до этого момента, предоставляют параметр с именем auth, который позволяет вам передавать свои учетные данные.

* Одним из примеров **API**, который требует аутентификации, является **Authenticated User API на GitHub**.
* Это конечная точка веб-сервиса, которая предоставляет информацию о профиле аутентифицированного пользователя.
* Чтобы отправить запрос **API-интерфейсу аутентифицированного пользователя**, вы можете передать свое имя пользователя и пароль на **GitHub** через кортеж в `get()`.

In [70]:
from getpass import getpass
requests.get('https://api.github.com/user', auth=('username', getpass()))

········


<Response [401]>

* Запрос выполнен успешно, если учетные данные, которые вы передали в кортеже `auth`, действительны. 
* Если вы попытаетесь сделать этот запрос без учетных данных, вы увидите, что код состояния **401 Unauthorized**.

* Когда вы передаете имя пользователя и пароль в кортеже параметру `auth`, вы используете учетные данные при помощи базовой схемы аутентификации **HTTP**.
* Таким образом, вы можете создать тот же запрос, передав подробные учетные данные базовой аутентификации, используя **HTTPBasicAuth**.

In [73]:

from requests.auth import HTTPBasicAuth
from getpass import getpass
requests.get(
            'https://api.github.com/user',
            auth=HTTPBasicAuth('username', getpass()))

········


<Response [401]>

* Хотя вам не нужно явно указывать обычную аутентификацию, может потребоваться аутентификация с использованием другого метода. 
* `Requests` предоставляет другие методы аутентификации, например: **HTTPDigestAuth** и **HTTPProxyAuth**.

* Вы даже можете предоставить свой собственный механизм аутентификации. 
* Для этого необходимо сначала создать подкласс **AuthBase**. 
* Затем происходит имплементация `__call__()`. (**Имплементация** - «осуществление, выполнение, практическая реализация»)

In [74]:
import requests
from requests.auth import AuthBase
 
class TokenAuth(AuthBase):
    """Implements a custom authentication scheme."""
 
    def __init__(self, token):
        self.token = token
 
    def __call__(self, r):
        """Attach an API token to a custom auth header."""
        r.headers['X-TokenAuth'] = f'{self.token}'  # Python 3.6+
        return r
 
 
requests.get('https://httpbin.org/get', auth=TokenAuth('12345abcde-token'))

<Response [200]>

* Здесь пользовательский механизм **TokenAuth** получает специальный токен. 
* Затем этот токен включается заголовок **X-TokenAuth** запроса.

* Плохие механизмы аутентификации могут привести к уязвимостям безопасности. 
* Поэтому, если службе по какой-то причине не нужен настраиваемый механизм аутентификации, вы всегда можете использовать проверенную схему аутентификации, такую как **Basic** или **OAuth**.

Далее рассмотрим использование `requests` в **SSL сертификатах**.

## 13. Python Requests проверка SSL сертификата

* Всякий раз, когда данные, которые вы пытаетесь отправить или получить, являются конфиденциальными, безопасность важна.
* Вы общаетесь с защищенными сайтами через HTTP, устанавливая зашифрованное соединение с использованием **SSL**, что означает, что проверка **SSL сертификата** целевого сервера имеет решающее значение.
* Хорошей новостью является то, что `requests` по умолчанию все делает сам. Однако в некоторых случаях необходимо внести определенные поправки.
* Если требуется отключить проверку **SSL-сертификата**, параметру `verify` функции запроса можно присвоить значение `False`.

In [75]:
 requests.get('https://api.github.com', verify=False)

C:\ProgramData\anaconda3\lib\site-packages\urllib3\connectionpool.py:1045: InsecureRequestWarning: Unverified HTTPS request is being made to host 'api.github.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(


<Response [200]>

* В случае небезопасного запроса `requests` предупреждает о возможности потери информации и просит сохранить данные или отказаться от запроса.

**Примечание:** Для предоставления сертификатов `requests` использует пакет, который вызывается `certifi`. Это дает понять `requests`, каким ответам можно доверять. Поэтому вам следует часто обновлять `certifi`, чтобы обеспечить максимальную безопасность ваших соединений.

## 14. Python Requests производительность приложений

* При использовании `requests`, особенно в среде приложений, важно учитывать влияние на производительность. 
* Такие функции, как контроль таймаута, сеансы и ограничения повторных попыток, могут помочь обеспечить бесперебойную работу приложения.

### `timeout` (Таймауты)

* Когда вы отправляете встроенный запрос во внешнюю службу, вашей системе нужно будет дождаться ответа, прежде чем двигаться дальше.
* Если ваше приложение слишком долго ожидает ответа, запросы к службе могут быть сохранены, пользовательский интерфейс может пострадать или фоновые задания могут зависнуть.
* По умолчанию в `requests` на ответ время не ограничено, и весь процесс может занять значительный промежуток.
* По этой причине вы всегда должны указывать время ожидания, чтобы такого не происходило. 
* Чтобы установить время ожидания запроса, используйте параметр `timeout`.
* `timeout` может быть целым числом или числом с плавающей точкой, представляющим количество секунд ожидания ответа до истечения времени ожидания.

In [81]:
requests.get('https://api.github.com', timeout=1)

<Response [200]>

In [82]:
requests.get('https://api.github.com', timeout=3.05)

<Response [200]>

* В первом примере запрос истекает через 1 секунду. Во втором примере запрос истекает через 3,05 секунды.

* Мы также можете передать **кортеж**. 
* Это – таймаут соединения (время, за которое клиент может установить соединение с сервером), а второй – таймаут чтения (время ожидания ответа, как только ваш клиент установил соединение):

In [83]:
requests.get('https://api.github.com', timeout=(2, 5))

<Response [200]>

* Если запрос устанавливает соединение в течение 2 секунд и получает данные в течение 5 секунд после установления соединения, то ответ будет возвращен, как это было раньше. 
* Если время ожидания истекло, функция вызовет исключение `Timeout`.



Наша программа может поймать `исключение Timeout` и ответить соответственно:

In [84]:
import requests
from requests.exceptions import Timeout
 
try:
    response = requests.get('https://api.github.com', timeout=1)
except Timeout:
    print('The request timed out')
else:
    print('The request did not time out')

The request did not time out


## 15. Объект `Session` в `Requests`

* До сих пор вы имели дело с **requests API высокого уровня**, такими как `get()` и `post()`.
* Эти функции являются абстракцией того, что происходит, когда вы делаете свои запросы. 
* Они скрывают детали реализации, такие как управление соединениями, так что вам не нужно о них беспокоиться.
* Под этими абстракциями находится **класс** под названием `Session`. 
* Если вам необходимо настроить контроль над выполнением запросов или повысить производительность ваших запросов, вам может потребоваться использовать `Session` напрямую.

> **Сессии** используются для сохранения параметров в запросах.

Если вы хотите использовать одну и ту же аутентификацию для нескольких запросов, вы можете использовать сеанс:

In [89]:
import requests
from getpass import getpass
 
# используя менеджер контента, можно убедиться, что ресурсы, применимые
# во время сессии будут свободны после использования
with requests.Session() as session:
    session.auth = ('username', getpass())
 
    # Instead of requests.get(), you'll use session.get()
    response = session.get('https://api.github.com/user')
 
# здесь можно изучить ответ 
print(response.headers)
print('===================================================================================================================')
print(response.json())

········
{'Server': 'GitHub.com', 'Date': 'Sun, 12 Mar 2023 17:05:47 GMT', 'Content-Type': 'application/json; charset=utf-8', 'Content-Length': '131', 'X-GitHub-Media-Type': 'github.v3; format=json', 'x-github-api-version-selected': '2022-11-28', 'X-RateLimit-Limit': '60', 'X-RateLimit-Remaining': '56', 'X-RateLimit-Reset': '1678644162', 'X-RateLimit-Used': '4', 'X-RateLimit-Resource': 'core', 'Access-Control-Expose-Headers': 'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=31536000; includeSubdomains; preload', 'X-Frame-Options': 'deny', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '0', 'Referrer-Policy': 'origin-when-cross-origin, strict-origin-when-cross-origin', '

* Каждый раз, когда вы делаете запрос `session`, после того как он был инициализирован с учетными данными аутентификации, учетные данные будут сохраняться.

* Первичная оптимизация производительности сеансов происходит в форме постоянных соединений. 
* Когда ваше приложение устанавливает соединение с сервером с помощью `Session`, оно сохраняет это соединение в пуле соединений.
* Когда ваше приложение снова хочет подключиться к тому же серверу, оно будет использовать соединение из пула, а не устанавливать новое.

# 16. `HTTPAdapter` — Максимальное количество повторов запроса в `Requests`

* В случае сбоя запроса возникает необходимость сделать повторный запрос. 
* Однако `requests` не будет делать это самостоятельно. 
* Для применения функции повторного запроса требуется реализовать собственный **транспортный адаптер**.



* **Транспортные адаптеры** позволяют определить набор конфигураций для каждой службы, с которой вы взаимодействуете.
* Предположим, вы хотите, чтобы все запросы к https://api.github.com были повторены три раза, прежде чем, наконец, появится **ConnectionError**. 
* Для этого нужно построить **транспортный адаптер**, установить его параметр **max_retries** и подключить его к существующему объекту `Session`.

In [91]:
import requests
from requests.adapters import HTTPAdapter
from requests.exceptions import ConnectionError
 
github_adapter = HTTPAdapter(max_retries=3)
 
session = requests.Session()
 
# использование `github_adapter` для всех запросов, которые начинаются с указанным URL
session.mount('https://api.github.com', github_adapter)
 
try:
    session.get('https://api.github.com')
except ConnectionError as ce:
    print(ce)

* При установке **HTTPAdapter**, **github_adapter** к `session`, `session` будет придерживаться своей конфигурации для каждого запроса к https://api.github.com.

* **Таймауты**, **транспортные адаптеры** и **сессии** предназначены для обеспечения эффективности используемого кода и стабильности приложения.

> *Данная статья является переводом статьи:* **Python’s Requests Library (Guide)** --->>>   https://realpython.com/python-requests/